google_api_lib/tasks.py

```diff
@@ -620,51 +620,61 @@ def update_active_users(self, hunt_id):
+                        Value("-"),
+                        F("puzzle_id"),
+                        output_field=CharField(),
+                    )
                 )
+                .filter(user_puzzle_pair__in=user_puzzle_pairs)
+                .select_for_update()
+            )
+            old_keys = [(e.user_id, e.puzzle_id) for e in old_activities]
+
+            PuzzleActivity.objects.bulk_create(
+                [
+                    PuzzleActivity(
+                        puzzle_id=puzzle_pk,
+                        user_id=user_pk,
+                        last_edit_time=last_edit_time,
+                        num_edits=num_edits,
+                    )
+                    for (
+                        (user_pk, puzzle_pk),
+                        (last_edit_time, num_edits),
+                    ) in latest.items()
+                    if (user_pk, puzzle_pk) not in old_keys
+                ],
             )
-            .filter(user_puzzle_pair__in=user_puzzle_pairs)
-            .select_for_update()
-        )
-        old_keys = [(e.user_id, e.puzzle_id) for e in old_activities]
-
-        PuzzleActivity.objects.bulk_create(
-            [
-                PuzzleActivity(
-                    puzzle_id=puzzle_pk,
-                    user_id=user_pk,
-                    last_edit_time=last_edit_time,
-                    num_edits=num_edits,
-                )
-                for (
-                    (user_pk, puzzle_pk),
-                    (last_edit_time, num_edits),
-                ) in latest.items()
-                if (user_pk, puzzle_pk) not in old_keys
-            ],
-        )
 
-        updates = []
-        for activity in old_activities:
-            (last_edit_time, num_edits) = latest[(activity.user_id, activity.puzzle_id)]
-            activity.last_edit_time = last_edit_time
-            activity.num_edits += num_edits
-            updates.append(activity)
+            updates = []
+            for activity in old_activities:
+                (last_edit_time, num_edits) = latest[
+                    (activity.user_id, activity.puzzle_id)
+                ]
+                activity.last_edit_time = last_edit_time
+                activity.num_edits += num_edits
+                updates.append(activity)
 
-        if updates:
-            PuzzleActivity.objects.bulk_update(
-                updates, fields=["last_edit_time", "num_edits"]
-            )
+            if updates:
+                PuzzleActivity.objects.bulk_update(
+                    updates, fields=["last_edit_time", "num_edits"]
+                )
 
-        hunt.last_active_users_update_time = end
-        hunt.save()
+            hunt.last_active_users_update_time = end
+            hunt.save()
+    except IntegrityError:
+        # attempt to avoid the error next time by clearing cache
+        logger.warning("IntegrityError while updating active users, clearing cache")
+        cache.clear()
```

Comment:
I'm not like totally sold this is a good idea but if you want to merge this as a kind of bandaid it's probably fine

Reply:
lol I agree, I think a better solution to last year's issue is to remove a puzzle's key from the cache when its deleted. I'll leave this unmerged for now and merge it in last minute if we don't do a better bandaid before Hunt.